### DISCUSSION TITLE
Is there something like join_any?

### DISCUSSION THREAD

**Discussion by lz1998** [tokio-rs/tokio#4708] [Q&A] [answered]
Is there something like `join_any`, which is similar to `try_join`.

I want to get the first `Ok(TcpStream)`.

If all failed, then return `Err`.

```rust
    async fn connect(&self, _: &Arc<Client>) -> std::io::Result<TcpStream> {
        tokio::select! {
            conn1 = TcpStream::connect(SocketAddr::new(Ipv4Addr::new(1,2,3,4).into(), 80)) => {
                return conn1 // TODO if error, not return
            }
            conn2 = TcpStream::connect(SocketAddr::new(Ipv4Addr::new(5,6,7,8).into(), 80)) => {
                return conn2 // TODO if error, not return
            }
        }
    }
```

**Comment by Darksonn** [comment, marked as answer]
Use a pattern to disable the `tokio::select!` branch if it returns an error.
```Rust
async fn connect(&self, _: &Arc<Client>) -> std::io::Result<TcpStream> {
    tokio::select! {
        Ok(conn1) = connect1() => Ok(conn1),
        Ok(conn2) = connect2() => Ok(conn2),
        else => Err(...)
    }
}
```

**Reply by lz1998** [reply]
Does `select` support vector? What should I do if there is a lot of addresses in a `Vec`?

**Reply by Darksonn** [reply]
No, but there are alternatives you can use for that situation. In Tokio, there is the [`JoinSet`][1] and in the futures crate there is [`FuturesUnordered`][2]. They can both be used to implement that.

[1]: https://docs.rs/tokio/latest/tokio/task/struct.JoinSet.html
[2]: https://docs.rs/futures/latest/futures/stream/struct.FuturesUnordered.html
